src/algebraic_geometry/elliptic_curve/weierstrass.lean

```diff
@@ -382,17 +382,164 @@ See Zulip thread:
+  cases ideal.mem_span_singleton'.mp (ideal.quotient.eq_zero_iff_mem.mp hx) with p hp,
+  apply_fun degree at hp,
+  rw [W.monic_polynomial.degree_mul, degree_polynomial, degree_C $ X_sub_C_ne_zero x] at hp,
+  cases p.degree; cases hp
+end
+
+/-- The class of the element $Y - y(X)$ in $R[W]$ for some $y(X) \in R[X]$. -/
+@[simp] noncomputable def Y_class : W.coordinate_ring := adjoin_root.mk W.polynomial $ X - C y
+
+lemma Y_class_ne_zero [nontrivial R] : Y_class W y ≠ 0 :=
+begin
+  intro hy,
+  cases ideal.mem_span_singleton'.mp (ideal.quotient.eq_zero_iff_mem.mp hy) with p hp,
+  apply_fun degree at hp,
+  rw [W.monic_polynomial.degree_mul, degree_polynomial, degree_X_sub_C] at hp,
+  cases p.degree; cases hp
+end
+
+/-- The ideal $\langle X - x \rangle$ of $R[W]$ for some $x \in R$. -/
+@[simp] noncomputable def X_ideal : ideal W.coordinate_ring := ideal.span {X_class W x}
+
+/-- The ideal $\langle Y - y(X) \rangle$ of $R[W]$ for some $y(X) \in R[X]$. -/
+@[simp] noncomputable def Y_ideal : ideal W.coordinate_ring := ideal.span {Y_class W y}
+
+variable {W}
+
+lemma smul (x : R[X]) (y : W.coordinate_ring) : x • y = adjoin_root.mk W.polynomial (C x) * y :=
+(algebra_map_smul W.coordinate_ring x y).symm
+
+lemma smul_basis_eq_zero [nontrivial R] {p q : R[X]}
+  (hpq : p • 1 + q • adjoin_root.mk W.polynomial X = 0) : p = 0 ∧ q = 0 :=
+begin
+  have h := fintype.linear_independent_iff.mp (coordinate_ring.basis W).linear_independent ![p, q],
+  erw [fin.sum_univ_succ, basis_zero, fin.sum_univ_one, basis_one] at h,
+  exact ⟨h hpq 0, h hpq 1⟩
+end
+
+lemma exists_smul_basis_eq [nontrivial R] (x : W.coordinate_ring) :
+  ∃ p q : R[X], p • 1 + q • adjoin_root.mk W.polynomial X = x :=
+begin
+  have h := (coordinate_ring.basis W).sum_equiv_fun x,
+  erw [fin.sum_univ_succ, fin.sum_univ_one, basis_zero, basis_one] at h,
+  exact ⟨_, _, h⟩
+end
+
+lemma norm_smul_basis [nontrivial R] (p q : R[X]) :
+  algebra.norm R[X] (p • 1 + q • adjoin_root.mk W.polynomial X)
+    = p ^ 2 - p * q * (C W.a₁ * X + C W.a₃)
+      - q ^ 2 * (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆) :=
+begin
+  have h : (p • 1 + q • adjoin_root.mk W.polynomial X) * adjoin_root.mk W.polynomial X
+    = (q * (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆)) • 1
+      + (p - q * (C W.a₁ * X + C W.a₃)) • adjoin_root.mk W.polynomial X,
+  { simp only [smul],
+    exact adjoin_root.mk_eq_mk.mpr
+      ⟨C q, by { simp only [weierstrass_curve.polynomial, C_sub, C_mul], ring1 }⟩ },
+  simp_rw [algebra.norm_eq_matrix_det W^.coordinate_ring.basis, matrix.det_fin_two,
+           algebra.left_mul_matrix_eq_repr_mul, basis_zero, mul_one, basis_one, h, map_add,
+           finsupp.add_apply, map_smul, finsupp.smul_apply, ← basis_zero, ← basis_one,
+           basis.repr_self_apply, if_pos, if_neg one_ne_zero, if_neg zero_ne_one, smul_eq_mul],
+  ring1
+end
+
+lemma nat_degree_norm_ne_one [nontrivial R] [no_zero_divisors R] (x : W.coordinate_ring) :
+  (algebra.norm R[X] x).nat_degree ≠ 1 :=
```

Comment:
I've been able to compute the degree exactly; it depends on a lemma and the main proof is indeed longer, but maybe you can golf it down.
```
lemma nat.with_bot.add_one_le_of_lt {n m : with_bot ℕ} (h : n < m) : n + 1 ≤ m :=
begin
  cases n, { exact bot_le },
  cases m, exacts [(not_lt_bot h).elim, with_bot.some_le_some.2 (with_bot.some_lt_some.1 h)],
end

lemma degree_norm_eq_max [is_domain R] {p q : R[X]} :
  degree (algebra.norm R[X] $ p • 1 + q • adjoin_root.mk W.polynomial X) =
  max (2 • degree p) (2 • degree q + 3) :=
begin
  have dq : degree (q ^ 2 * (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆)) = 2 • q.degree + 3,
  { rw [degree_mul, degree_pow, ← one_mul (X^3), ← C_1, degree_cubic (@one_ne_zero R _ _ _)] },
  have dpq : degree (p * q * (C W.a₁ * X + C W.a₃)) ≤ p.degree + q.degree + 1,
  { rw [degree_mul, degree_mul],
    apply add_le_add_left ((degree_add_le _ _).trans $ max_le_iff.2 _),
    exact ⟨degree_C_mul_X_le _, degree_C_lt.le⟩ },
  rw [norm_smul_basis],
  obtain rfl | hq := eq_or_ne q 0,
  { simp_rw [degree_zero, mul_zero, zero_pow zero_lt_two, zero_mul, sub_zero, degree_pow],
    exact (max_bot_right _).symm },
  have hqb : 2 • q.degree ≠ ⊥,
  { rw [two_nsmul, with_bot.add_ne_bot], split; exact mt degree_eq_bot.1 hq },
  cases le_or_lt p.degree (q.degree + 1),
  { have h' := (nsmul_le_nsmul_of_le_right h 2),
    rw [nsmul_add, nsmul_one] at h',
    convert degree_sub_eq_right_of_degree_lt _; rw dq,
    { exact max_eq_right_iff.2 (h'.trans $ add_le_add_left dec_trivial _) },
    apply (degree_sub_le _ _).trans_lt (max_lt_iff.2 ⟨_, dpq.trans_lt _⟩),
    { rw [degree_pow], exact h'.trans_lt (with_bot.add_lt_add_left hqb dec_trivial) },
    { rw [add_assoc],
      refine (add_le_add_right h _).trans_lt _,
      rw [add_add_add_comm, ← two_nsmul],
      exact with_bot.add_lt_add_left hqb dec_trivial } },
  { have h' := nat.with_bot.add_one_le_of_lt h,
    rw [add_assoc] at h',
    have h'' := (nsmul_le_nsmul_of_le_right h' 2),
    rw [nsmul_add] at h'',
    rw sub_sub, convert degree_sub_eq_left_of_degree_lt _; rw degree_pow,
    { exact max_eq_left_iff.2 ((add_le_add_left dec_trivial _).trans h'') },
    refine (degree_add_le _ _).trans_lt (max_lt_iff.2 ⟨dpq.trans_lt _, _⟩),
    { rw [two_nsmul, add_assoc],
      exact with_bot.add_lt_add_left (λ hpb, (not_lt_bot $ hpb.subst h).elim) h },
    { rw dq, exact lt_of_lt_of_le (with_bot.add_lt_add_left hqb dec_trivial) h'' } },
end

lemma degree_norm_ne_one [is_domain R] (x : W.coordinate_ring) :
  (algebra.norm R[X] x).degree ≠ 1 :=
λ h, begin
  rcases exists_smul_basis_eq x with ⟨p, q, rfl⟩,
  rw degree_norm_eq_max at h,
  cases q.degree,
  { cases p.degree with dp, {cases h}, cases dp, {cases h}, cases dp; cases h },
  { exact (lt_max_of_lt_right dec_trivial).ne' h },
end

lemma nat_degree_norm_ne_one [is_domain R] (x : W.coordinate_ring) :
  (algebra.norm R[X] x).nat_degree ≠ 1 :=
mt (degree_eq_iff_nat_degree_eq_of_pos zero_lt_one).2 $ degree_norm_ne_one x
```

Reply:
Thanks - I was hesitating whether to prove `degree` or `nat_degree` first. Ultimately I chose to do `nat_degree` first, otherwise I would have needed such a lemma, but the existing `data/nat/with_bot` file is already synchronised with mathlib4, and I didn't want to make the mathlib4 PR because I didn't know any Lean4. If you can help me do the corresponding mathlib4 PR I'll switch to this proof.

Reply:
I opened #18174 for nat.with_bot and #18160 for removing [nontrivial] instances. I feel like this degree computation belongs to the theory of Newton polygons: the degree is the valuation (pole order) at infinity, and because the middle term lies below the convex hull of the two outer terms, it can never contribute to the max.

Reply:
I've golfed your proof down to 20 lines using `linarith`, which is now shorter than my original code! Along the way I somehow removed the need for `nat.with_bot.add_one_le_of_lt` so we can remove the dependency, but I think having it in the first place is useful anyway. I don't quite understand how the PR removing `nontrivial` instances affects this PR though.

Reply:
Wow that's impressive! Glad you removed the dependency. Your proof of `nat_degree_norm_ne_one` from the `degree` version is longer than the one I suggested, though. #18160 would allow you define `basis` as (untested)
```
(subsingleton_or_nontrivial R).by_cases
  (λ h, by exactI basis.reindex (adjoin_root.power_basis' W.monic_polynomial).basis $ fin_congr W.nat_degree_polynomial)
  (λ h, by exactI default)
```
where `default` comes from the `unique` instance added in #18160. Then it's possible to remove `[nontrivial R]` assumptions from all theorems about `basis` in this PR; due to uniqueness the `default` basis must be equal to `{1, X}`, but I guess `nontriviality R` can already solve the subsingleton case most of the time.

Note that my #18177 has also been merged, which is intended to be used in the proof of `norm_smul_basis`. I realize it would be better if I made a version that specializes to `adjoin_root.mk X` ... The `power_basis` isn't directly applicable because we're working with `basis` and I didn't add a way to construct `power_basis` in the `subsingleton` case in #18160. So some more work would need to be done to streamline everything; maybe that should be deferred to a later PR instead of blocking this one.

Reply:
I removed `nontrivial`, but I think there should be a better way to prove `basis_zero` and `basis_one` - it's horrendous now.